Documentation/architecture.rst

```diff
@@ -0,0 +1,155 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    http://docs.cilium.io
+
+.. _arch_guide:
+
+############
+Architecture
+############
+
+This document describes the Cilium architecture. It focuses on
+documenting the BPF datapath hooks to implement the Cilium datapath, how
+the Cilium datapath integrates with the container orchestration layer, and the
+objects shared between the layers e.g. the BPF datapath and Cilium agent.
+
+Datapath
+============
+
+The Linux kernel supports a set of BPF hooks in the networking stack
+that can be used to run BPF programs. The Cilium datapath uses these
+hooks to load BPF programs that when used together create higher level
+networking constructs.
+
+The following is a list of the hooks used by a Cilium and a brief
+description. For a more thorough documentation on specifics of each
+hook see :ref:`bpf_guide`.
+
+* **XDP:** The XDP BPF hook is at the earliest point possible in the networking driver
+  and triggers a run of the BPF program upon packet reception. This
+  achieves the best possible packet processing performance since the
+  program runs directly on the packet data before any other processing
+  can happen. This hook is ideal for running filtering programs that
+  drop malicious or unexpected traffic and other common DDOS protection
+  mechanisms.
+
+* **Traffic Control Ingress:** BPF programs attached to the traffic control (tc)
+  ingress hook are attached to a networking interface, same as XDP, but
+  will run after the networking stack has done initial processing
+  of the packet. The hook is run before the L3 layer of the stack but has access
```

Comment:
maybe add: initial processing of the packet such as GRO?

Reply:
hmm seems a bit arbitrary to pick on GRO to me. Other processing includes vlan tag extraction, various other skb fields, etc.